### PR TITLE
[distill] how to get feature map 

### PR DESCRIPTION
**Pull request by ceci3**
how to get feature map for distill

### CODE REVIEW

File: paddleslim/dygraph/dist/distill.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+                final_distill_dict['teacher'][feature_type + '_' + str(
+                    sidx) + '_' + str(tidx)] = self.tea_outs_dict[
+                        feature_type + '_' + str(tidx)]
+        return final_distill_dict
+
+    def forward(self, *inputs, **kwargs):
+        stu_batch_outs = self._student_models.forward(*inputs, **kwargs)
+        tea_batch_outs = self._teacher_models.forward(*inputs, **kwargs)
+        distill_inputs = self.post_outputs()
+        ### batch is None just for now
```
Comment (minghaoBD) [inline paddleslim/dygraph/dist/distill.py:183]: 请问这个batch 是什么物理含义呢？batch_size吗？

File: tests/dygraph/test_distill.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                    _logger.info(
+                        "Test | step {}: acc1 = {:}, acc5 = {:}".format(
+                            batch_id, acc_top1.numpy(), acc_top5.numpy()))
+
+            _logger.info("Test |Average: acc_top1 {}, acc_top5 {}".format(
+                np.mean(avg_acc[0]), np.mean(avg_acc[1])))
+
+        def train(model):
+            adam = paddle.optimizer.Adam(
+                learning_rate=0.001, parameters=distill_model.parameters())
```
Comment (qingqing01) [inline tests/dygraph/test_distill.py]: 68行接口参数名字叫model，不是distill_model，下面都是在用distill_model

File: paddleslim/dygraph/dist/distill.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+DistillConfig = namedtuple(
+    'DistillConfig',
+    [
+        ### list(dict): config of each mapping layers.
+        'layers_config',
+    ])
+DistillConfig.__new__.__defaults__ = (None, ) * len(DistillConfig._fields)
+
+
+def transpose_config(config):
```
Comment (qingqing01) [inline paddleslim/dygraph/dist/distill.py]: 这里config类型是 DistillConfig吗？

用户要填充的参数，需要作为接口。 s_feature_idx， t_feature_idx范围是啥，feature_type如何设置，loss_function候选是啥等。 当前没有接口/数据结构 告诉用户。

File: paddleslim/dygraph/dist/distill.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                    int(config['s_feature_idx']), int(config['t_feature_idx'])
+                ])
+
+    def prepare_loss(self):
+        self.distill_loss = CombinedLoss(self._loss_config_list)
+
+    def prepare_outputs(self):
+        stu_outs_dict = collections.OrderedDict()
+        tea_outs_dict = collections.OrderedDict()
+        stu_outs_dict = self._adaptors_S(self._student_models, stu_outs_dict)
```
Comment (qingqing01) [inline paddleslim/dygraph/dist/distill.py]: _adaptors_S这里实现接口有一定的约束。 需要抽象基类，对接口做约束

File: paddleslim/dygraph/dist/distill.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        tea_outs_dict = self._adaptors_T(self._teacher_models, tea_outs_dict)
+        return stu_outs_dict, tea_outs_dict
+
+    def post_outputs(self):
```
Comment (qingqing01) [inline paddleslim/dygraph/dist/distill.py]: 不暴露给用户的内部函数，_开头

**Review by qingqing01 — COMMENTED** [review]
add comments

File: tests/dygraph/test_distill.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            mode='test', backend='cv2', transform=transform)
+
+        place = paddle.CUDAPlace(0) if paddle.is_compiled_with_cuda(
+        ) else paddle.CPUPlace()
+        train_reader = paddle.io.DataLoader(
+            train_dataset, drop_last=True, places=place, batch_size=64)
+        test_reader = paddle.io.DataLoader(
+            val_dataset, places=place, batch_size=64)
+
+        def adaptor(model, mapping_layers):
```
Comment (wanghaoshuang) [inline tests/dygraph/test_distill.py]: 最好以Class的方式实现`adaptor`.
在`AdaptorBase`中实现方法：
- add_hooks: 根据用户添加的mapping信息注册hook

在`MobileNetV1Adaptor `中实现方法：
- mapping_layers: 添加['hidden', 'logits']与layer name 之间的mapping信息

构造`MobileNetV1Adaptor `时，调用`mapping_layers`，
构造`Distill`对象时，调用` add_hooks`, 根据用户的配置，添加部分hooks.

File: paddleslim/dygraph/dist/distill.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+            loss_config[str(c['loss_function'])]['key'] = c[
+                'feature_type'] + '_' + str(c['s_feature_idx']) + '_' + str(c[
+                    't_feature_idx'])
+            ### TODO: support list of student models and teacher_models
+            loss_config[str(c['loss_function'])][
+                'model_name_pairs'] = [['student', 'teacher']]
+            self._loss_config_list.append(loss_config)
+        self.prepare_loss()
+
+    def get_distill_idx(self):
```
Comment (qingqing01) [inline paddleslim/dygraph/dist/distill.py]: 下面是set/init行为，没有返回，不是get

File: tests/dygraph/test_distill.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+        def train(model):
+            adam = paddle.optimizer.Adam(
+                learning_rate=0.001, parameters=distill_model.parameters())
+
+            for batch_id, data in enumerate(train_reader):
+                img = paddle.to_tensor(data[0])
+                label = paddle.to_tensor(data[1])
+                student_out, teacher_out, distill_loss = distill_model(img)
+                loss = paddle.nn.functional.loss.cross_entropy(student_out,
```
Comment (wanghaoshuang) [inline tests/dygraph/test_distill.py:101]: 可以考虑把这个loss也放到Distill对象中管理。

File: paddleslim/dygraph/dist/distill.py (hunk 1 above)
Comment (ceci3) [reply paddleslim/dygraph/dist/distill.py:183]: 指的是一个batch的数据，再难样本蒸馏的方向里计算loss的时候会需要一些类间信息，就先留着了

File: tests/dygraph/test_distill.py (hunk 2 above)
Comment (ceci3) [reply tests/dygraph/test_distill.py]: done, thanks

**Review by wanghaoshuang — APPROVED** [review]
